Mos: Genercize SourceID logic

duytnguyendtn opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our code of conduct,
https://github.com/spacetelescope/jdaviz/blob/main/CODE_OF_CONDUCT.md . -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable viz component(s). -->

This PR addresses the original request of updating our Mosviz Generic Metadata parser to attempt to use the SOURCEID header value, with the additional stretch goal of unifying the NIRISS Source Name extraction logic with the generic parser's

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->


### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [ ] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [ ] Is a milestone set? Milestone is only currently required for PRs related to Imviz MVP.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

pllim commented on jdaviz/configs/mosviz/plugins/parsers.py:
```diff
@@ -463,14 +463,28 @@ def mos_meta_parser(app, data_obj, ids=None, spectra=False, sp1d=False):
+    except KeyError:
+        if filepaths:
+            # remove leading path to file name
+            filenames = [os.path.basename(d) for d in filepaths]
```
This might also work (untested).
```suggestion
            filenames = list(map(os.path.basename, filepaths))
```

duytnguyendtn replied:
Stale

pllim commented on jdaviz/configs/mosviz/plugins/parsers.py:
```diff
@@ -463,14 +463,28 @@ def mos_meta_parser(app, data_obj, ids=None, spectra=False, sp1d=False):
+
+    Parameters
+    ----------
+    hdus : a list of HDUs (NOT an HDUList!) to check the header of. Filtering should
+        be done in advance to decide which HDUs to check against (e.g. Sci only)
+    filepaths : Optional. A list of filepaths to fallback on if no header values are
+        identified
```
```suggestion
    hdus : list of HDU
        A list of HDUs (NOT an HDUList!) to check the header of. Filtering should
        be done in advance to decide which HDUs to check against (e.g., SCI only).
    filepaths : list of str, optional
        A list of filepaths to fallback on if no header values are identified.
```

duytnguyendtn commented:
Thanks for the review @pllim! I spent some time architecting a more performant algorithm by using `hdu.get()` rather than relying on `try except`. 

One thing I do want to get feedback on is the bare except. In this situation, I toyed with the principle that `source_id` lookup should NEVER stop the application from doing its job (i.e. Even if there is a catastrophic failure, we should fall back on the `FALLBACK_NAME` we define earlier, rather than fail outright). What do other devs think?

pllim commented:
Re: bare except -- is `except Exception` not enough for your case? Bare except might also intercept things outside of Python like Cntrl + C (if I remember correctly).

duytnguyendtn commented:
Ahh, I didn't know `except Exception` had different behavior than a bare except! Good to know, and I think it would be good for this usecase :)

duytnguyendtn commented on jdaviz/configs/mosviz/plugins/parsers.py:
```diff
@@ -463,14 +463,28 @@ def mos_meta_parser(app, data_obj, ids=None, spectra=False, sp1d=False):
+    # Try to extract the Source ID:
+    try:
+        src_names = ["Source ID: " + str(x.header['SOURCEID']) for x in hdus]
+    except KeyError:
```
Reworked except logic, take a look! :)

kecnry reviewed (commented):
What is the desired behavior for the [NIRISS example notebook](https://github.com/spacetelescope/jdaviz/blob/main/notebooks/MosvizNIRISSExample.ipynb)?  This seems to replace the previous "Source Catalog: F150W Source ID: N" with just numbers 1-10 for that example.

javerbukh reviewed (commented):
I agree with @kecnry about using more readable names for the NIRISS notebook and type checking for `header_keys` but otherwise this looks great! If possible it would be nice to add some tests here but the code looks good to me.

duytnguyendtn commented on jdaviz/configs/mosviz/plugins/parsers.py:
```diff
@@ -463,14 +464,51 @@ def mos_meta_parser(app, data_obj, ids=None, spectra=False, sp1d=False):
+    """
+    src_names = list()
+    # If the user only provided one key to search, put it in a list for the upcoming for loop
+    if type(header_keys) != list:
```
Thanks! Implemented a better typecheck on header_keys to check if its Iterable, but exclude dictionaries and strings. 

As for checking if a header key is unsupported, I'm going to raise a quick test I did that slightly complicates things...
```
>>> from astropy.io.fits import PrimaryHDU
>>> x = PrimaryHDU()
>>> x.header[True] = "test" 
>>> x.header.get(True)
'test'
```
Obviously it makes the most sense to type check for strings... but astropy doesn't seem to enforce that???? Oddly enough, it doesn't show in the header:
```
>>> x.header
SIMPLE  =                    T / conforms to FITS standard
BITPIX  = 'test    '           / array data type
NAXIS   =                    0 / number of array dimensions
EXTEND  = 
```
I went and tried to see how far I could go, and tried a spectrum1D as the key, but that didn't work
```
>>> x.header[spec1d] = "test"
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "E:\STScI\gitRepos\jdaviz\envmain\lib\site-packages\astropy\io\fits\header.py", line 208, in __setitem__
    self._update((key, value, comment))
  File "E:\STScI\gitRepos\jdaviz\envmain\lib\site-packages\astropy\io\fits\header.py", line 1673, in _update
    keyword = keyword.strip().upper()
AttributeError: 'Spectrum1D' object has no attribute 'strip'
```
So my question here is... what's unsupported?

duytnguyendtn commented:
> What is the desired behavior for the [NIRISS example notebook](https://github.com/spacetelescope/jdaviz/blob/main/notebooks/MosvizNIRISSExample.ipynb)? This seems to replace the previous "Source Catalog: F150W Source ID: N" with just numbers 1-10 for that example.

I'm going to ping the POs on this one (@PatrickOgle @camipacifici @orifox @eteq) to see what we want here. To reiterate the goal of this effort, this PR is trying to take the step towards a unified loader by having both the NIRISS and generic parser pull the source names from the same place. Hence I removed the old source name formatting and replaced it with just what's returned from the header. 
https://github.com/spacetelescope/jdaviz/blob/main/jdaviz/configs/mosviz/plugins/parsers.py#L615-L616
The "Source ID" part is, upon reflection, redundant since it's in the column name. The actual additional information that was added was the filter name. What do the POs think should be returned here?

PatrickOgle commented:
@duytnguyendtn The source ID number seems plenty to me since the table header should give the necessary context.

rosteen commented on jdaviz/configs/mosviz/plugins/parsers.py:
```diff
@@ -463,14 +465,51 @@ def mos_meta_parser(app, data_obj, ids=None, spectra=False, sp1d=False):
+        lookup)
+    """
+    src_names = list()
+    # If the user only provided one key to search, put it in a list for the upcoming for loop
+    if isinstance(header_keys, Iterable) and not isinstance(header_keys, (str, dict)):
```
This comment doesn't seem to match the functionality of the code - the comment seems to imply that if it's not iterable it should be put in a list e.g. `[header_keys]`, but it doesn't do anything in that case. Is this doing everything that was intended?

duytnguyendtn replied:
You are 100% correct! This is entirely a bug. I've pushed a fix and clarified a comment. This slipped by because I forgot to write a test for it when the feature was suggested, so I wrote a few tests to capture it. (Proud of me Pey Lian? 😄)

rosteen commented on jdaviz/configs/mosviz/plugins/parsers.py:
```diff
@@ -667,7 +706,7 @@ def mos_niriss_parser(app, data_dir, obs_label=""):
         # We then populate the table inside this context manager as _add_to_table
         # does operations that also trigger link manager updates.
-        _add_to_table(app, source_ids, "Source ID")
+        _add_to_table(app, source_ids, "Source Name")
```
It bothers me a little that this is now "Name" instead of "ID" when the contents of the column might be simple integers (which seem more like IDs than names to me). Maybe there could be some code to decide between "Name"/"ID"? That could get messy internally though if other code refers to the column...personally I would default to "ID".

duytnguyendtn replied:
Here's the dilemma I was faced with. I agree with you Ricky, that really what we're after here is some sort of "Identifier" that *identifies* the target, whatever that is, a word, a number, etc. The problem is `SOURCEID` is an ACTUAL KEYWORD used by the JWST pipeline, so I don't want to put `Source ID` because I don't want it to be confused with the actual keyword itself!

We could perhaps have it be "Target ID" or something, but I'm trying to avoid using the same terminology as the actual header keyword we're defaulting to. Maybe this is a PO design question (@PatrickOgle because you brought up the original idea, but also @camipacifici @orifox @larrybradley). Regardless, I think it should be the same for all parsers (regardless of instrument), and would be nicer for it to be uniform (i.e. NOT switch between Name/ID) if at all possible.

rosteen replied:
> The problem is `SOURCEID` is an ACTUAL KEYWORD used by the JWST pipeline, so I don't want to put `Source ID` because I don't want it to be confused with the actual keyword itself!

I think it makes sense to have the `SOURCEID` value go in a column called `Source ID`, we should get some other opinions on whether other people find that confusing.

rosteen replied:
@camipacifici @PatrickOgle @orifox any opinion here?

pllim replied:
Crazy ideas:

1. Have the column name be dynamic. If you are grabbing from "Source ID", that column will be "Source ID"; If you are grabbing from "TARGNAME", that column will be "TARGNAME". However, doing it like this, you cannot load two datasets with different headers.
2. Have one additional column that tells you which header keyword that source name is taken from. Name this new column something generic. This bloats the table a little.

rosteen reviewed (commented):
I left a couple minor comments, but it mostly looks good.

Jenneh commented:
I don't fully understand the context of this label, but is this an area where we can use a hover tool tip to more fully explain regardless of which term we use? It could help since the term will probably not be fully obvious to all users.

duytnguyendtn commented:
@Jenneh This column is meant to represent the "Name" of the target in Mosviz. In an ideal world, it would be things like "Andromeda Galaxy", "Orion Nebula", "NGC3329". However, as Ricky points out, sometimes the titles people can give their targets could be something as simple as numbers (i.e. Source "1", "2", "3", etc.). So the question is what should the column that holds this information be called? The origin of this PR is that sometimes Mosviz shows "Source Name" as the column title. Other times, it's "Source ID". I settled on "Source Name" as the unifying column title because `SOURCEID` is an actual FITS keyword generated by the JWST pipeline. Is there a better column title you can think of that might be more useful?

Jenneh commented:
Is there a reason the column shouldn't be called "Target"? 
We use that terminology is search forms and I think astronomers understand what it means. In some cases we use "object" but that has a slightly different connotation.

rosteen commented:
@duytnguyendtn Would you rebase this so RTD passes? Otherwise it seems good to go. IIRC the PO's were happy with "Identifier", so I'll approve and merge once tests pass.

duytnguyendtn commented:
Thanks @rosteen! Missed that failing doc test. Rebased. Let's see that the tests pass this time
